Frameworks/CoreText/CTLine.mm

```diff
@@ -283,16 +283,44 @@ CGRect CTLineGetImageBounds(CTLineRef line, CGContextRef context) {
 }
 
 /**
- @Status Caveat
- @Notes Values only reflect the typographical maximums for the font - string is only evaulated
-        for width
+ @Status Interoperable
+ @Notes
 */
 double CTLineGetTypographicBounds(CTLineRef lineRef, CGFloat* ascent, CGFloat* descent, CGFloat* leading) {
-    if (!lineRef) {
+    _CTLine* line = static_cast<_CTLine*>(lineRef);
+
+    if (!line || [line->_runs count] == 0) {
         return 0;
     }
 
-    _CTLine* line = static_cast<_CTLine*>(lineRef);
+    // Created with impossible values FLT_MIN which signify they need to be populated
+    if (line->_ascent == FLT_MIN || line->_descent == FLT_MIN) {
+        for (_CTRun* run in static_cast<id<NSFastEnumeration>>(line->_runs)) {
+            DWRITE_GLYPH_METRICS glyphMetrics[run->_dwriteGlyphRun.glyphCount];
+            THROW_IF_FAILED(run->_dwriteGlyphRun.fontFace->GetDesignGlyphMetrics(run->_dwriteGlyphRun.glyphIndices,
+                                                                                 run->_dwriteGlyphRun.glyphCount,
+                                                                                 glyphMetrics,
+                                                                                 run->_dwriteGlyphRun.isSideways));
+
+            CTFontRef font = static_cast<CTFontRef>([run->_attributes objectForKey:static_cast<NSString*>(kCTFontAttributeName)]);
+            CGFloat pointSize = kCTFontSystemFontSize;
+            if (font) {
+                pointSize = CTFontGetSize(font);
+            }
+
+            DWRITE_FONT_METRICS fontMetrics;
+            run->_dwriteGlyphRun.fontFace->GetMetrics(&fontMetrics);
+            CGFloat pointsPerDesignUnit = pointSize / fontMetrics.designUnitsPerEm;
```

Comment:
Raj's comment was made at a time when the division result was already outside the function, so you might as well have cached it. In general I have the following issues with _CoreTextScaleMetric:

- The name is not very informative. For CTFont, __CTFontScaleMetric was more forgivable since the context was smaller, and the scope of what it could be was smaller. _CoreTextScaleMetric on the other hand is in DWriteWrapper, and could be any number of things. A better name might be something like _CoreTextDesignMetricToPoints()
- Changing the name to that though, makes it pretty obvious that the function just hides a multiplication, in which case it feels a bit silly to have a dedicated helper for it?
- It feels extra silly to have a dedicated helper for it and also assume that the caller will have the knowledge to do this division outside, rather than shifting the burden to the function.
- In general, caching that result is a pretty small optimization, and we'd prefer easier-to-read/maintain code over the small gains that caching would get us. I'm fine with leaving this function the way it is if you think this is the most readable state, but I don't want optimization to be your foremost reason.

---
In reply to: [80529943](https://github.com/Microsoft/WinObjC/pull/1027#discussion_r80529943) [](ancestors = 80529943,80528495)